Multiple proposal improvements to the API

hartbit opened the pull request:
* Remove `Logging` and move its functions to `Logger`. `make` becomes an initializer.
* Introduce a argument-less initializer to `Logger` to initialize it with a global level when people want to use a global logger.
* Remove all factory initializers in example and test code.

hartbit committed:
Multiple proposal improvements to the API

* Move `Logging.make` to a `Logger` initializer.
* Remove all factory initializers in example and test code.

tomerd commented:
@hartbit thank you, lots of good ideas here. we plan to address all feedback next week and finalize the proposal

weissi commented on Sources/Logging/Logging.swift:
```diff
@@ -96,25 +96,46 @@ public struct Logger {
 
-// This is the logging system itself, it's mostly used to obtain loggers and to set the type of the `LogHandler`
-// implementation.
-public enum Logging {
```
@hartbit how would you feel about instead renaming `Logging` to `LoggingSystem`? Previously we went for `Logging` because `Logging.make(...)` was such an important function. But after your PR, `Logging.make(...)` would be replaced by `Logger(...)`. So the importance of `Logging` itself is vastly reduced.

I feel like we should rename it to `LoggingSystem` and keep the `bootstrap` and `make` (which maybe should be `makeLogger`) methods. In most programs, the type `LoggingSystem` will only be used once: to bootstrap the logging system with

    LoggingSystem.bootstrap(...)

and I think that reads well.

hartbit replied:
Sounds like a nice compromise

weissi commented on Sources/Logging/Logging.swift:
```diff
@@ -96,25 +96,46 @@ public struct Logger {
+        self = logHandlerFactoryLock.withReaderLock { Logger(logHandlerFactory(label)) }
+    }
+
+    public init() {
```
can we remove the initialiser without label?

hartbit replied:
Do you think the convenience dangerously pushes developers towards the one-global logger?

weissi replied:
I think it’s quite useful to have the logger at least identify themselves with some form of ID string. And I think the burden to the user passing a string there is not too high. But offering one without the need of passing an ID string is just too easy so I fear that no one will end up using the ID strings and they quickly become useless.

Sure, the users could just pass the empty string etc but I don’t think they would.

It’s basically like `DispatchQueue(label: “foo.bar”)` and people seem to fill in more or less sensible stuff.

hartbit replied:
Makes sense. Will make the necessary changes

weissi reviewed (changes requested):
thanks @hartbit , this looks mostly good, left some comments though.

weissi commented:
@hartbit any thoughts on this?

hartbit commented:
@weissi I've modified the PR to put bootstrap back on `Logging`. I decided to leave its name to `Logging` for now, we can always rename it in a separate PR. I have also removed the label-less `Logger` initialiser.

weissi reviewed (approved):
Thanks @hartbit! That lgtm. There are conflicts though, if you fix them we can merge this.

hartbit commented:
@weissi Rebased, fixed conflicts, and rewrote commit message to reflect the new changes.
